### PR TITLE
fix: change default multiLineStart pattern to handle stackoverflow error

### PR DESCRIPTION
**Pull request by junfuchen99**
**Issue #, if available:**

**Description of changes:**
Change default `multiLineStart` pattern to check leading whitespaces directly.
Add handling of `StackOverflow` error when matching `multiLineStart` pattern .
Add testing.

**Why is this change necessary:**

**How was this change tested:**
unit test.

**Any additional information or context required to review the change:**

**Checklist:**
 - [ ] Updated the README if applicable
 - [ ] Updated or added new unit tests
 - [ ] Updated or added new integration tests
 - [ ] Updated or added new end-to-end tests

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: src/main/java/com/aws/greengrass/logmanager/CloudWatchAttemptLogsProcessor.java [hunk 1]
```diff
@@ -399,6 +401,26 @@ private void updateCloudWatchAttemptLogInformation(String fileName,
...
+     * @param logLine                      The processed log line
+     * @return Whether the log line is the start of a new log event.
+     */
+    private boolean checkLogStartPattern(final Pattern multiLineStartPattern, final String logLine) {
+        if (multiLineStartPattern != null) {
+            try {
+                return multiLineStartPattern.matcher(logLine).find();
+            } catch (StackOverflowError e) {
+                logger.atWarn().cause(e).log("Unable to match log against pattern {}. Checking for trailing "
+                                + "whitespaces instead. Detailed log: {}", multiLineStartPattern.pattern(), logLine);
```
Comment (MikeDombo) [inline src/main/java/com/aws/greengrass/logmanager/CloudWatchAttemptLogsProcessor.java:417]: the log line is already super long causing an overflow, I don't think you want to log out the problematic log line here.

File: src/main/java/com/aws/greengrass/logmanager/CloudWatchAttemptLogsProcessor.java [hunk 2]
```diff
@@ -399,6 +401,26 @@ private void updateCloudWatchAttemptLogInformation(String fileName,
...
+     * @param multiLineStartPattern        The pattern for start of new log event
+     * @param logLine                      The processed log line
+     * @return Whether the log line is the start of a new log event.
+     */
+    private boolean checkLogStartPattern(final Pattern multiLineStartPattern, final String logLine) {
+        if (multiLineStartPattern != null) {
+            try {
+                return multiLineStartPattern.matcher(logLine).find();
+            } catch (StackOverflowError e) {
+                logger.atWarn().cause(e).log("Unable to match log against pattern {}. Checking for trailing "
```
Comment (MikeDombo) [inline src/main/java/com/aws/greengrass/logmanager/CloudWatchAttemptLogsProcessor.java:416]: you aren't checking for trailing whitespace though, you look at the beginning.

Comment (junfuchen99) [reply src/main/java/com/aws/greengrass/logmanager/CloudWatchAttemptLogsProcessor.java:416]: yea type

File: src/test/java/com/aws/greengrass/logmanager/CloudWatchAttemptLogsProcessorTest.java [hunk 3]
```diff
@@ -543,6 +534,91 @@ void GIVEN_unstructured_log_WHEN_breaches_event_size_and_batch_size_limits_THEN_
     }
 
     @Test
+    void GIVEN_component_multiline_pattern_default_WHEN_logs_start_with_whitespace_THEN_append_to_previous_log() throws IOException{
```
Comment (MikeDombo) [inline src/test/java/com/aws/greengrass/logmanager/CloudWatchAttemptLogsProcessorTest.java:537]: do any of these tests cause the stack overflow? We need a regression test to make sure it is handled correctly

Comment (junfuchen99) [reply src/test/java/com/aws/greengrass/logmanager/CloudWatchAttemptLogsProcessorTest.java:540]: added a test

**Review by MikeDombo — APPROVED** [review]

File: src/test/java/com/aws/greengrass/logmanager/CloudWatchAttemptLogsProcessorTest.java [hunk 4]
```diff
@@ -542,6 +536,135 @@ void GIVEN_unstructured_log_WHEN_breaches_event_size_and_batch_size_limits_THEN_
...
+            assertNotNull(logEventsForStream1.getLogEvents());
+            assertEquals(4, logEventsForStream1.getLogEvents().size());
+        } finally {
+            assertTrue(file.delete());
+        }
+    }
+
+    @SuppressWarnings("PMD.CloseResource")
+    @Test
+    void GIVEN_component_multiline_pattern_set_WHEN_log_lines_trigger_stack_overflow_THEN_use_default_pattern_and_()
```
Comment (shaguptashaikh) [inline src/test/java/com/aws/greengrass/logmanager/CloudWatchAttemptLogsProcessorTest.java:626]: nit - name of this test seems incomplete `_use_default_pattern_and_`

File: src/test/java/com/aws/greengrass/logmanager/CloudWatchAttemptLogsProcessorTest.java [hunk 5]
```diff
@@ -542,6 +536,135 @@ void GIVEN_unstructured_log_WHEN_breaches_event_size_and_batch_size_limits_THEN_
...
+                    .next();
+            assertNotNull(logEventsForStream1.getLogEvents());
+            assertEquals(4, logEventsForStream1.getLogEvents().size());
+        } finally {
+            assertTrue(file.delete());
+        }
+    }
+
+    @Test
+    void GIVEN_component_multiline_pattern_set_WHEN_log_lines_do_not_match_patternTHEN_append_to_previous_log() throws IOException{
```
Comment (shaguptashaikh) [inline src/test/java/com/aws/greengrass/logmanager/CloudWatchAttemptLogsProcessorTest.java:581]: Nit- formatting in test name

**Review by shaguptashaikh — APPROVED** [review]

**Review by tilo-chen — DISMISSED** [review]

**Review by tilo-chen — APPROVED** [review]

**Review by MikeDombo — APPROVED** [review]

**Pull request merged**